Pal/lib/hex.h

```diff
@@ -0,0 +1,52 @@
+/* -*- mode:c; c-file-style:"k&r"; c-basic-offset: 4; tab-width:4; indent-tabs-mode:nil; mode:auto-fill; fill-column:78; -*- */
+/* vim: set ts=4 sw=4 et tw=78 fo=cqt wm=0: */
+
+/* Copyright (C) 2014 OSCAR lab, Stony Brook University
+   2017 University of North Carolina at Chapel Hill
+   This file is part of Graphene Library OS.
+
+   Graphene Library OS is free software: you can redistribute it and/or
+   modify it under the terms of the GNU General Public License
+   as published by the Free Software Foundation, either version 3 of the
+   License, or (at your option) any later version.
+
+   Graphene Library OS is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */
+
+#ifndef HEX_H
+#define HEX_H
+
+/* This function is a helper for debug printing. 
+ * It accepts a pointer to a numerical value, and 
+ * formats it as a hex string, for printing.
+ * size is the number of bytes pointed to by hex.
+ * str is the caller-provided buffer, len is the length of the buffer.
+ * The len must be at least (size * 2)+1.
+ * 
+ * Note that it does not normalize for endianness, and pads to the
+ * size the compiler things the string is.
+ */
+#include <assert.h>
+static inline __attribute__((always_inline))
+char * __hex2str(void * hex, int size, char *str, int len)
+{
+    static char * ch = "0123456789abcdef";
+    assert(len >= size * 2 + 1);
+
+    for (int i = 0 ; i < size ; i++) {
+        unsigned char h = ((unsigned char *) hex)[i];
+        str[i * 2] = ch[h / 16];
+        str[i * 2 + 1] = ch[h % 16];
+    }
+
+    str[size * 2] = 0;
+    return str;
+}
+
+#define hex2str(array, str, len) __hex2str(array, sizeof(*array), str, len)
```

Comment:
Why this version prints only the first element of the array?
Also, IMO it'd be better to ensure that `array` is in fact an array, not a plain pointer.

Reply:
`array` is a macro argument, so it should be wrapped it parentheses, otherwise a call like:
```C
hex2str(array_begin+5, out, out_len)
```
Will quietly lead to very unexpected results (i.e. calculating `sizeof(*array_begin+5)`).

Reply:
Agreed.  Most actual uses are for printing a hash or a mac, where it is some number of bytes.  I'm not sure how to check this without some fancier typed wrappers.

Reply:
The problem is that for `char arr[10]` `sizeof(*arr)` is 1, not 10, so it will print only the first byte. So, it should probably be `sizeof(arr)` + a static assert that `arr` is in fact an array. For a solution for the latter you can look e.g. here: https://stackoverflow.com/questions/16794900/validate-an-argument-is-array-type-in-c-c-pre-processing-macro-on-compile-time.

Reply:
How is the latest push?

Reply:
Almost good ;) You should wrap `arg` in both `IS_ARRAY` and `IS_INDEXABLE` in parentheses, otherwise a similar problem like the one mentioned before may happen.